tests/Aspire.Hosting.MongoDB.Tests/MongoDbFunctionalTests.cs

```diff
@@ -0,0 +1,341 @@
+        var pipeline = new ResiliencePipelineBuilder()
+            .AddRetry(new() { MaxRetryAttempts = 10, Delay = TimeSpan.FromSeconds(1) })
+            .Build();
+
+        string? volumeName = null;
+        string? bindMountPath = null;
+
+        try
+        {
+            var builder1 = CreateDistributedApplicationBuilder();
+            var mongodb1 = builder1.AddMongoDB("mongodb");
+            var db1 = mongodb1.AddDatabase(dbName);
+
+            if (useVolume)
+            {
+                // Use a deterministic volume name to prevent them from exhausting the machines if deletion fails
+                volumeName = VolumeNameGenerator.CreateVolumeName(mongodb1, nameof(WithDataShouldPersistStateBetweenUsages));
+
+                // if the volume already exists (because of a crashing previous run), try to delete it
+                DockerUtils.AttemptDeleteDockerVolume(volumeName);
+                mongodb1.WithDataVolume(volumeName);
+            }
+            else
+            {
+                bindMountPath = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+                mongodb1.WithDataBindMount(bindMountPath);
+            }
+
+            using (var app = builder1.Build())
+            {
+                await app.StartAsync();
+                try
+                {
+                    var hb = Host.CreateApplicationBuilder();
+
+                    hb.Configuration.AddInMemoryCollection(new Dictionary<string, string?>
+                    {
+                        [$"ConnectionStrings:{db1.Resource.Name}"] = await db1.Resource.ConnectionStringExpression.GetValueAsync(default)
+                    });
+
+                    hb.AddMongoDBClient(db1.Resource.Name);
+
+                    using (var host = hb.Build())
+                    {
+                        await host.StartAsync();
+
+                        await pipeline.ExecuteAsync(async token =>
+                        {
+                            var mongoDatabase = host.Services.GetRequiredService<IMongoDatabase>();
+                            await CreateTestDataAsync(mongoDatabase, token);
+                        }, cts.Token);
+                    }
+                }
+                finally
+                {
+                    // Stops the container, or the Volume/mount would still be in use
+                    await app.StopAsync();
+                }
+            }
+
+            var builder2 = CreateDistributedApplicationBuilder();
+            var mongodb2 = builder2.AddMongoDB("mongodb");
+            var db2 = mongodb2.AddDatabase(dbName);
+
+            if (useVolume)
+            {
+                mongodb2.WithDataVolume(volumeName);
+            }
+            else
+            {
+                //mongodb shutdown has delay,so without delay to running instance using same data and second instance failed to start.
+                await Task.Delay(TimeSpan.FromSeconds(10));
```

Comment:
We need to make this more deterministic. Waiting for X seconds is usually how we get random/flakey failures.

Reply:
What's your suggestion?
I don't have any idea.

Reply:
Try to connect to the database until it fails?

Reply:
Is there any helper method exists to check container is fully stopped?

Reply:
I was expecting something like https://github.com/dotnet/aspire/issues/4878. If you want to prototype trying to make that work. cc @karolz-ms @DamianEdwards

Reply:
@sebastienros - should we log an issue for this wait case, and then we can merge this PR?

Reply:
Filed https://github.com/dotnet/aspire/issues/5199